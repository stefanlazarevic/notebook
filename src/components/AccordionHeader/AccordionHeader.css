.AccordionHeader {
    box-sizing: border-box;

    position: relative;

    width: 100%;

    padding: 5px 10px;

    background-color: #f8f9fb;

    border: 1px solid #e4e4e4;

    text-align: left;

    cursor: pointer;
}

.AccordionHeader:focus {
    outline: 2px solid #a06b9a;
}