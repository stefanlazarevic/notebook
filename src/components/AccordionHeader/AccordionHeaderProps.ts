import PropTypes, { InferProps, Validator } from "prop-types";

export const AccordionHeaderPropTypes = {
	/**
	 *
	 */
	id: PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	testid: PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	index: PropTypes.number as Validator<number | undefined>,
	/**
	 * @default false
	 */
	expanded: PropTypes.bool as Validator<boolean | undefined>,
	/**
	 *
	 */
	controls: PropTypes.string as Validator<string | undefined>,
	/**
	 * @default false
	 */
	disabled: PropTypes.bool as Validator<boolean | undefined>,
	/**
	 *
	 */
	children: PropTypes.node as Validator<React.ReactNode | React.ReactNodeArray | undefined>,
	/**
	 * @default 2
	 */
	level: PropTypes.oneOf([1, 2, 3, 4, 5, 6]) as Validator<1 | 2 | 3 | 4 | 5 | 6 | undefined>,
	/**
	 *
	 */
	className: PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	onFocus: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onBlur: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onClick: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onArrowDown: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onArrowUp: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onHome: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
	/**
	 *
	 */
	onEnd: PropTypes.func as Validator<(event: React.SyntheticEvent, index?: number) => void | undefined>,
};

export type AccordionHeaderProps = InferProps<typeof AccordionHeaderPropTypes>;
