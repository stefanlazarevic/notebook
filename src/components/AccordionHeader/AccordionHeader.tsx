import React from "react";

import "./AccordionHeader.css";

import { AccordionHeaderProps, AccordionHeaderPropTypes } from "./AccordionHeaderProps";

import useClassNames from "../Utils/hooks/classNames";

/**
 * Компонента која представља наслов секције који служи за контролисање
 * видљивости секције.
 *
 * @see https://www.w3.org/TR/wai-aria-practices/#accordion
 */
function AccordionHeader(props: AccordionHeaderProps) {
	const className = useClassNames("AccordionHeader", props.className);

	/**
	 *
	 * @param event
	 */
	function onClick(event: React.MouseEvent<HTMLButtonElement>) {
		if (typeof props.onClick === "function" && !props.disabled) {
			props.onClick(event, props.index);
		}
	}

	/**
	 *
	 * @param event
	 */
	function onKeyDown(event: React.KeyboardEvent<HTMLButtonElement>) {
		if (props.disabled) {
			return;
		}

		const { keyCode } = event;

		if ((keyCode === 32 || keyCode === 13) && typeof props.onClick === "function") {
			event.preventDefault();

			props.onClick(event, props.index);
		}

		if (keyCode === 40 && typeof props.onArrowDown === "function") {
			props.onArrowDown(event, props.index);
		}

		if (keyCode === 38 && typeof props.onArrowUp === "function") {
			props.onArrowUp(event, props.index);
		}

		if (keyCode === 36 && typeof props.onHome === "function") {
			props.onHome(event, props.index);
		}

		if (keyCode === 35 && typeof props.onEnd === "function") {
			props.onEnd(event, props.index);
		}
	}

	return (
		<div role="heading" aria-level={props.level}>
			<button
				id={props.id}
				data-testid={props.testid}
				data-index={props.index}
				tabIndex={0}
				className={className}
				aria-expanded={props.expanded}
				aria-controls={props.controls}
				disabled={props.disabled}
				onKeyDown={onKeyDown}
				onClick={onClick}
				onFocus={props.onFocus}
				onBlur={props.onBlur}
			>
				{props.children}
			</button>
		</div>
	);
}

AccordionHeader.propTypes = AccordionHeaderPropTypes;

AccordionHeader.defaultProps = {
	expanded: false,
	disabled: false,
	level: 2,
};

AccordionHeader.displayName = "AccordionHeader";

export default AccordionHeader;
