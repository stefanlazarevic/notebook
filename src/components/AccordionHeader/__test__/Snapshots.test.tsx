import React from "react";
import { render, cleanup } from "@testing-library/react";
import AccordionHeader from "../index";

afterEach(cleanup);

it("Подразумеванa својства се подударају са последњим снимком.", () => {
	const { asFragment } = render(<AccordionHeader />);

	expect(asFragment()).toMatchSnapshot();
});

it("Сва подешена својства се подударају са последњим снимком.", () => {
	const { asFragment } = render(<AccordionHeader id="id" testid="testid" className="CustomAccordionHeader" />);

	expect(asFragment()).toMatchSnapshot();
});
