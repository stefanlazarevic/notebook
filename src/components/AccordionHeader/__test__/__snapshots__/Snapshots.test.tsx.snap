// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Подразумеванa својства се подударају са последњим снимком. 1`] = `
<DocumentFragment>
  <div
    aria-level="2"
    role="heading"
  >
    <button
      aria-expanded="false"
      class="AccordionHeader"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Сва подешена својства се подударају са последњим снимком. 1`] = `
<DocumentFragment>
  <div
    aria-level="2"
    role="heading"
  >
    <button
      aria-expanded="false"
      class="AccordionHeader CustomAccordionHeader"
      data-testid="testid"
      id="id"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
